core/types/block_ext_test.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestCopyHeader(t *testing.T) {
+	t.Run("empty_header", func(t *testing.T) {
+		empty := &Header{}
+
+		headerExtra := &HeaderExtra{}
+		extras.Header.Set(empty, headerExtra)
+
+		cpy := CopyHeader(empty)
+
+		want := &Header{
+			Difficulty: new(big.Int),
+			Number:     new(big.Int),
+		}
+
+		headerExtra = &HeaderExtra{}
+		extras.Header.Set(want, headerExtra)
+
+		assert.Equal(t, want, cpy)
+	})
+
+	t.Run("filled_header", func(t *testing.T) {
+		h := &Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		headerExtra := &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+		extras.Header.Set(h, headerExtra)
+
+		allFieldsAreSet(t, h)
+		allFieldsAreSet(t, headerExtra)
+
+		cpy := CopyHeader(h)
+
+		want := &Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		headerExtra = &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+		extras.Header.Set(want, headerExtra)
+
+		assert.Equal(t, want, cpy)
+
+		// Mutate each non-value field to ensure they are not shared
+		fieldsAreDeepCopied(t, h, cpy)
+		fieldsAreDeepCopied(t, GetHeaderExtra(h), GetHeaderExtra(cpy))
+	})
+}
+
+func ptrTo[T any](x T) *T { return &x }
+
+func allFieldsAreSet(t *testing.T, x any) {
+	t.Helper()
+	require.Equal(t, reflect.Ptr.String(), reflect.TypeOf(x).Kind().String(), "x must be a pointer")
+	v := reflect.ValueOf(x).Elem()
+	typ := v.Type()
+	require.Equal(t, reflect.Struct.String(), typ.Kind().String())
```

Comment:
```suggestion
	require.Equal(t, reflect.Struct, typ.Kind())
```

Reply:
Actually that's on purpose, since otherwise it just prints a confusing integer (enum). It's much better to print out the string to know what's wrong.